### PR TITLE
Added deeplink for selection form elements

### LINKED ISSUE #912: Link SmartContent and InternalLinks entries to edit mask
It would be nice if the entries in the smart content (and then also for media, internal links, snippets) are deep links, so that a click leads to the corresponding edit form.

image

/cc @berndhep WDYT?

### PR DESCRIPTION
**Pull request by wachterjohannes**
__tasks:__

- [ ] gather feedback for my changes
- [x] Media select
- [x] Smart content
- [x] Internal link
- [x] Snippet
- [x] Contact
- [x] Account
- [x] Styles
- [x] Smart-Content and Internal-Links (uuid of index page ....)

__informations:__

| q                | a
| ---------------- | ---
| Fixed tickets    | fixes #912
| Linked PR | https://github.com/massiveart/husky/pull/612
| BC breaks        | not really BC but for data-provider there is a new config option
| Documentation PR | https://github.com/sulu-io/sulu-docs/pull/183

### CODE REVIEW

**Comment by chirimoya** [comment]
@wachterjohannes please finish this PR soonish.

**Comment by wachterjohannes** [comment]
i have looked at this a few weeks ago ... its not that easy because this is
implemented on the old system therefor i dont think its easy to "upgrade"
...
Am 28.10.2015 11:26 schrieb "Thomas Schedler" <notifications@github.com>:

**Comment by webmozart** [comment]
Is there any progress on this issue? Should we close this?

**Comment by wachterjohannes** [comment]
let me take a look at this this week maybe i could rebase it. it would be a quite cool feature.

**Comment by wachterjohannes** [comment]
@webmozart this feature is rebased and works quite well. i will finish it for 1.2 only styles and the snippet-content-type is missing

**Comment by webmozart** [comment]
@wachterjohannes Thanks! :)

File: UPGRADE.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
...
+
+The smart-content is now able to deep-link selected elements. To provide proper-links add the deep-links to the
+DataProvider configuration.
+
+```php
+$this->configuration = Builder::create()
+    ...
+    ->setDeepLink('content/contents/{webspace}/{locale}/edit:{id}/details')
+    ->getConfiguration();
+```
```
Comment (danrot) [inline UPGRADE.md]: I think we've agreed on not putting new features into the `UPGRADE.md`? Probably this PR is older than this decision :smiley:

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js [hunk 2]
```diff
@@ -592,10 +608,17 @@ define(['services/husky/util'], function(util) {
                 var ul = this.sandbox.dom.createElement('<ul class="' + constants.contentListClass + '"/>');
 
                 this.sandbox.util.foreach(this.items, function(item, index) {
-                    this.sandbox.dom.append(ul, _.template(templates.contentItem)({
+                    var template = templates.contentItem;
+                    if (this.options.deepLink !== '') {
+                        template = templates.contentItemLink
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js]: `;` missing

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig [hunk 3]
```diff
@@ -11,13 +11,15 @@
...
      data-aura-display-options='{{ sulu_parameter_to_key_value(params.display_options.value) | json_encode }}'
-     {% if params.has.datasource %}
+{% if params.has.datasource %}
      data-aura-datasource='{{ params.datasource | json_encode }}'
-     {% endif %}
+{% endif %}
      {% if params.has.categories %}
      data-aura-category-root='{{ params.category_root.value }}'
-     {% endif %}
+{% endif %}
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: Why did you change the indentation here? If you've got a good reason please also change it for the ìf` corresponding to this `endif`.

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig [hunk 4]
```diff
@@ -11,13 +11,15 @@
...
      data-aura-url="/admin/api/items?{% if webspaceKey != '' %}webspace={{ webspaceKey }}&{% endif %}locale={{ languageCode }}&excluded=<%= options.id %>&provider={{ params.provider.value }}"
+     data-aura-navigate-event="sulu.content.navigate"
      data-aura-tags-auto-complete-url="/admin/api/tags?flat=true&sortBy=name&searchFields=name"
      data-aura-sort-by='{{ sulu_parameter_to_select(params.sorting.value, userLocale) | json_encode }}'
-     data-aura-present-as='{{ sulu_parameter_to_select(params.present_as.value, userLocale) | json_encode }}'></div>
+     data-aura-present-as='{{ sulu_parameter_to_select(params.present_as.value, userLocale) | json_encode }}'
+     data-aura-deep-link="{{ params['deep-link'].value }}"></div>
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: Any reason we are not using `deep_link` here, so that we don't need the array notation?

File: src/Sulu/Component/SmartContent/ContentType.php [hunk 5]
```diff
@@ -203,6 +203,7 @@ public function getDefaultParams(PropertyInterface $property = null)
                 'presentAs' => $configuration->hasPresentAs(),
             ],
             'datasource' => $configuration->getDatasource(),
+            'deep-link' => new PropertyParameter('deep-link', $configuration->getDeepLink()),
```
Comment (danrot) [inline src/Sulu/Component/SmartContent/ContentType.php]: Actually this line is connected to the comment in the twig template above. When doing a quick search on `PropertyParameter` I find examples written with underscores, dashes and in camel case. We should agree on one (and it should not be the dash, since it is unnecessarily complicated to handle in twig).

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig (hunk 3 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: because twig ignores such lines. if you indent the spaces will be in the output also.

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: hmm i will change it because two lines above its underscored

File: src/Sulu/Component/SmartContent/ContentType.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Component/SmartContent/ContentType.php]: i would go with "_" i will change it

Comment (wachterjohannes) [inline src/Sulu/Component/SmartContent/ContentType.php]: fixed

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: fixed

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig (hunk 3 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: indention fixed

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js (hunk 2 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js]: fixed

File: UPGRADE.md (hunk 1 above)
Comment (wachterjohannes) [inline UPGRADE.md]: fixed

**Comment by wachterjohannes** [comment]
@danrot fixed all comments and rebased this feature.
